auth-service.js

```diff
@@ -0,0 +1,500 @@
+ */
+const puppeteer = require("puppeteer");
+const axios = require("axios");
+const { v4: uuidv4 } = require("uuid");
+const crypto = require("crypto");
+
+let lastRequest; // Keep track of the last request for debugging
+let lastDebugMessage; // Keep a record of the last debug message
+let args; // Keep a reference to args for helper functions
+
+/**
+ * Executes auth code flow using provided arguments.
+ * @param argv yargs arguments.
+ * @returns {Promise<object>}
+ */
+function main(argv) {
+  args = argv;
+  return getAuthCode()
+    .then((payload) => {
+      if (payload.error) {
+        return payload;
+      }
+      if (argv["code-only"]) {
+        return { code: payload.code };
+      }
+      return getTokens(payload);
+    })
+    .then((response) => {
+      if (response.data) {
+        return response.data;
+      }
+      return response;
+    });
+}
+
+/**
+ * Use puppeteer to retrieve an authorization code.
+ * This mimicks the user login flow in a browser and is
+ * subject to breaking if the login page or form change.
+ * @returns {Promise<object>} the authorization code.
+ */
+async function getAuthCode() {
+  const launchArgs = { args: [`--headless=${!args.show}`] };
+  if (!args.sandbox) {
+    launchArgs.args.push("--no-sandbox");
+  }
+  let authUrl;
+  let codeVerifier;
+  if (args["pkce"]) {
+    var codeChallenge = generateCodeChallenge();
+    authUrl = buildAuthorizeUrlForPKCE(
+      args["authorization-url"],
+      args["client-id"],
+      args["scope"],
+      args["redirect-uri"],
+      codeChallenge.code_challenge
+    );
+    codeVerifier = codeChallenge.code_verifier;
+  } else {
+    authUrl = `${args["authorization-url"]}/authorization?client_id=${
+      args["client-id"]
+    }&scope=${args["scope"].join(" ")}&response_type=code&redirect_uri=${
+      args["redirect-uri"]
+    }&aud=default&state=${uuidv4()}`;
+  }
+
+  if (args["launch"]) {
+    authUrl = `${authUrl}&launch=${args["launch"]}`;
+  }
+
+  let finalRedirectRequest;
+
+  debug("Launching browser");
+  const browser = await puppeteer.launch(launchArgs);
+  let page = await browser.newPage();
+
+  /*
+   * Intercept each request.
+   *
+   * Log requests; nest static/xhr requests.
+   *
+   * If URL contains "error", report it.
+   */
+  await page.setRequestInterception(true);
+  page.on("request", (interceptedRequest) => {
+    if (interceptedRequest.resourceType() === "document") {
+      lastRequest = interceptedRequest;
+      debug(`-- ${interceptedRequest.url()}`, false);
+    } else {
+      debug(`-- -- ${interceptedRequest.url()}`, false);
+    }
+
+    /*
+     * If we ever end up at a URL with an "error" param, report it.
+     *
+     * Example: invalid scope
+     */
+    if (getUrlParam(interceptedRequest.url(), "error")) {
+      return getError(
+        getUrlParam(interceptedRequest.url(), "error"),
+        getUrlParam(interceptedRequest.url(), "error_description")
+      );
```

Comment:
`browser.close()` will also need to be called anytime an error is returned

```suggestion
      const error = getError(
        getUrlParam(interceptedRequest.url(), "error"),
        getUrlParam(interceptedRequest.url(), "error_description")
      );
      browser.close();
      return error;
```